### PR TITLE
[SPARK-44839][SS][CONNECT] Better Error Logging when user tries to serialize spark session

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

Add a new error with detailed message when a user tries to access spark session and dataframe created using local spark session, in streaming spark connect `foreach`, `foreachBatch` and `StreamingQueryListener`.

Update: per reviewer's request, added a new error class `PySparkPicklingError`. Also move `UDTF_SERIALIZATION_ERROR` to the new class 

### Why are the changes needed?

Better error logging for the breaking change introduced in streaming spark connect.

### Does this PR introduce _any_ user-facing change?

Yes, before users can only see this non-informative error when they access a local spark session in their streaming connect related functions:
```
Traceback (most recent call last):
  File "/home/wei.liu/oss-spark/python/pyspark/serializers.py", line 459, in dumps
    return cloudpickle.dumps(obj, pickle_protocol)
  File "/home/wei.liu/oss-spark/python/pyspark/cloudpickle/cloudpickle_fast.py", line 73, in dumps
    cp.dump(obj)
  File "/home/wei.liu/oss-spark/python/pyspark/cloudpickle/cloudpickle_fast.py", line 632, in dump
    return Pickler.dump(self, obj)
TypeError: cannot pickle '_thread._local' object

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/wei.liu/oss-spark/python/pyspark/sql/connect/streaming/readwriter.py", line 508, in foreachBatch
    self._write_proto.foreach_batch.python_function.command = CloudPickleSerializer().dumps(
  File "/home/wei.liu/oss-spark/python/pyspark/serializers.py", line 469, in dumps
    raise pickle.PicklingError(msg)
_pickle.PicklingError: Could not serialize object: TypeError: cannot pickle '_thread._local' object
```

Now it is replaced with:
```
pyspark.errors.exceptions.base.PySparkPicklingError: [STREAMING_CONNECT_SERIALIZATION_ERROR] Cannot serialize the function `foreachBatch`. If you accessed the spark session, or a dataframe defined outside of the function, please be aware that they are not allowed in Spark Connect. For foreachBatch, please access the spark session using `df.sparkSession`, where `df` is the first parameter in your foreachBatch function. For StreamingQueryListener, please access the spark session using `self.spark`
```

### How was this patch tested?

Add unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by WweiL** [comment]
cc @rangadi @bogao007

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/connect/streaming/query.py [hunk 1]
```diff
@@ -237,7 +239,13 @@ def addListener(self, listener: StreamingQueryListener) -> None:
         listener._init_listener_id()
         cmd = pb2.StreamingQueryManagerCommand()
         expr = proto.PythonUDF()
-        expr.command = CloudPickleSerializer().dumps(listener)
+        try:
+            expr.command = CloudPickleSerializer().dumps(listener)
+        except pickle.PicklingError:
+            raise PySparkRuntimeError(
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/streaming/query.py:245]: @itholic do we need a dedicated error class for `PicklingError`? e.g., `PySparkPicklingError`?

File: python/pyspark/errors/error_classes.py [hunk 2]
```diff
@@ -718,6 +718,11 @@
       "pandas iterator UDF should exhaust the input iterator."
     ]
   },
+  "STREAMING_CONNECT_SERIALIZATION_ERROR" : {
+    "message" : [
+      "Cannot serialize the function `<name>`. If you accessed the spark session, or a dataframe defined outside of the function, or any object that contains a spark session, please be aware that they are not allowed in Spark Connect. For foreachBatch, please access the spark session using `df.sparkSession`, where `df` is the first parameter in your foreachBatch function. For StreamingQueryListener, please access the spark session using `self.spark`. For details please check out the PySpark doc for foreachBatch and StreamingQueryListener."
```
Comment (itholic) [inline python/pyspark/errors/error_classes.py:723]: nit: 
```suggestion
      "Cannot serialize the function `<name>`. If you accessed the Spark session, or a DataFrame defined outside of the function, or any object that contains a Spark session, please be aware that they are not allowed in Spark Connect. For `foreachBatch`, please access the Spark session using `df.sparkSession`, where `df` is the first parameter in your `foreachBatch` function. For `StreamingQueryListener`, please access the Spark session using `self.spark`. For details please check out the PySpark doc for `foreachBatch` and `StreamingQueryListener`."
```

File: python/pyspark/sql/connect/streaming/query.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/sql/connect/streaming/query.py:245]: Yes, I believe we need a new error class for new type of user-facing errors. Could you add a new PySpark error class for representing `pickle.PicklingError`?? See https://github.com/apache/spark/pull/40938/files as an example. I think we can also do it as a follow ups.

**Review by itholic — APPROVED** [review]
Otherwise looks good

**Comment by WweiL** [comment]
CC @ueshin, I also changed the serialization error thrown by UDTF, ptal!

**Comment by ueshin** [comment]
cc @allisonwang-db

**Review by allisonwang-db — APPROVED** [review]
The UDTF change looks good to me!

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

File: python/pyspark/sql/connect/streaming/query.py (hunk 1 above)
Comment (itholic) [reply python/pyspark/sql/connect/streaming/query.py]: > define a new PySparkPicklingError and replace this PySparkRuntimeError with that right?

Correct :-)

**Comment by WweiL** [comment]
@ueshin Can we have this merged? Thank you!

**Comment by HyukjinKwon** [comment]
@WweiL seems like your PR conflicted with this PR :-). Should probably update.

**Pull request closed without merging**